server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1252,8 +1254,34 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
                 }
             });
 
+        final StepListener<RepositoryData> filterRepositoryDataStep = new StepListener<>();
+
         // Step 2: Write new index-N blob to repository and update index.latest
         setPendingStep.whenComplete(newGen -> threadPool().executor(ThreadPool.Names.SNAPSHOT).execute(ActionRunnable.wrap(listener, l -> {
+            // BwC logic: Load snapshot version information if any snapshot is missing a version in RepositoryData so that the new
+            // RepositoryData contains a version for every snapshot
+            final List<SnapshotId> snapshotIdsWithoutVersion = repositoryData.getSnapshotIds().stream().filter(
```

Comment:
It is somewhat annoying that we have to load all `SnapshotInfo` here one-off but I think it's still the best solution available. 
Ideally, I was hoping to be able to just work from the assumption that if no version is set for a snapshot then it must be from before `8.0`/`7.6`. But I gave up on that idea since it breaks as soon as some older version cluster (for whatever reason), takes a snapshot and removes all the `version` fields when writing out a new `index-N` (maybe that's the wrong trade off though, glad to hear opinions on this :) ... obviously you could argue that loading all the `SnapshotInfo` is too high a price to pay just so the repo can move to the new metadata version earlier, but IMO even on S3 etc. loading 100 snapshots or so won't take all that long or cost any meaningful amount of $$$).